src/wagtail_ai/static_src/field_panel/main.ts

```diff
@@ -26,19 +14,296 @@ interface PreviewContent {
+  static handlers: Record<string, ContextHandler> = {};
+  static register(name: string, handler: ContextHandler) {
+    this.handlers[name] = handler;
+  }
+
+  static async get(controller: FieldPanelController) {
+    const context: Record<string, string> = {};
+    await Promise.allSettled(
+      Object.entries(ContextProvider.handlers).map(async ([key, handler]) => {
+        if (!controller.activePrompt?.prompt.includes(`{${key}}`)) return;
+        const result = await handler.call(controller);
+        if (![null, undefined].includes(result)) context[key] = result;
+      }),
+    );
+    return context;
+  }
+}
+
+type InputType = HTMLInputElement | HTMLTextAreaElement | HTMLDivElement;
+
+class FieldPanelController extends Controller<HTMLTemplateElement> {
+  static targets = ['dropdown', 'prompt', 'suggestion'];
+  static values = {
+    activePromptId: { type: String, default: '' },
+    dropdownTemplate: {
+      type: String,
+      default: '#wai-field-panel-dropdown-template',
+    },
+    imageInput: { type: String, default: '' },
+    mainInput: { type: String, default: '' },
+    prompts: { type: Array, default: [] },
+    state: { type: String, default: FieldPanelState.IDLE },
+    suggestion: { type: String, default: '' },
+  };
+  declare hasDropdownTarget: boolean;
+  declare dropdownTarget: HTMLDivElement;
+  declare promptTargets: HTMLButtonElement[];
+  declare hasSuggestionTarget: boolean;
+  declare suggestionTarget: HTMLDivElement;
+  declare activePromptIdValue: string;
+  declare dropdownTemplateValue: string;
+  declare imageInputValue: string;
+  declare mainInputValue: string;
+  declare promptsValue: DefaultPrompt[];
+  declare stateValue: FieldPanelState;
+  declare suggestionValue: string;
+  declare filteredPrompts: Prompt[];
+  declare fieldInput: HTMLElement;
+  declare input: InputType;
+  activePrompt: Prompt | null = null;
+  abortController: AbortController | null = null;
+  imageInput: HTMLInputElement | null = null;
+  dropdownController: DropdownController | null = null;
+
+  static defaultHandlers = {
+    async content_html(this: FieldPanelController) {
+      return (await this.getPreviewContent())?.innerHTML;
+    },
+    async content_text(this: FieldPanelController) {
+      return (await this.getPreviewContent())?.innerText;
+    },
+    async form_context_before(this: FieldPanelController) {
+      return this.formContext.before;
+    },
+    async form_context_after(this: FieldPanelController) {
+      return this.formContext.after;
+    },
+    async image_id(this: FieldPanelController) {
+      return this.imageInput?.value;
+    },
+    async input(this: FieldPanelController) {
+      return this.inputValue;
+    },
+    async max_length(this: FieldPanelController) {
+      return this.input.getAttribute('maxlength');
+    },
+  };
+
+  static {
+    Object.entries(this.defaultHandlers).forEach(([name, handler]) => {
+      ContextProvider.register(name, handler);
+    });
+  }
```

Comment:
This is the client-side equivalent of the validators. You can register your own function to provide the placeholder's values that need to be accessed from the client side, e.g. `content_text` for the preview content's text